[serve] Fail on the change of 'proxy_location' or 'http_options' parameters for the 'serve' API

axreldable opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. --> 

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Serve `http_options` and `proxy_location` (which determines `http_options.location`) can't be updated in runtime. When users try to update them they receive warnings in the following format:
```
WARNING 2025-11-03 20:51:10,462 serve 23061 -- The new client HTTP config differs from the existing one in the following fields: ['host', 'port', 'location']. The new HTTP config is ignored.
--
2025-11-03 20:51:59,590 WARNING serve_head.py:281 -- Serve is already running on this Ray cluster and it's not possible to update its HTTP options without restarting it. Following options are attempted to be updated: ['host'].
```

This PR:
- change warning to failing with the error `RayServeConfigException`
- eliminate `validate_http_options` function in the `dashboard.modules.serve.serve_head` as it's not needed anymore
- update documentation for [Proxy config](https://docs.ray.io/en/latest/serve/production-guide/config.html#proxy-config) that the parameter is global and can't be updated at runtime.
- change `HTTPOptions.location` default value `HeadOnly` -> `EveryNode` (it's likely the desired value)

--------------------------------------------------------------
User scenario:
- have a file `hello_world.py`:
```
# hello_world.py
from ray.serve import deployment

@deployment
async def hello_world():
    return "Hello, world!"

hello_world_app = hello_world.bind()
```
- execute commands:
```
ray stop
ray start --head
serve build -o config.yaml hello_world:hello_world_app  # generate `config.yaml`
serve deploy config.yaml
# in the `config.yaml` file update:
# proxy_location: EveryNode -> HeadOnly
# http_options.host: 0.0.0.0 -> 0.0.0.1
# http_options.port: 8000 -> 8001
serve deploy config.yaml
```
Output before the change:
```
# stdout:
bash$ serve deploy config.yaml
2025-09-14 17:19:15,606 INFO scripts.py:239 -- Deploying from config file: 'config.yaml'.
2025-09-14 17:19:15,619 SUCC scripts.py:359 -- 
Sent deploy request successfully.
 * Use `serve status` to check applications' statuses.
 * Use `serve config` to see the current application config(s).

# /tmp/ray/session_latest/logs/dashboard_ServeHead.log
2025-09-14 17:19:15,615 WARNING serve_head.py:177 -- Serve is already running on this Ray cluster and it's not possible to update its HTTP options without restarting it. Following options are attempted to be updated: ['location', 'host', 'port'].
```
Output after the change:
```
# stdout:
bash$ serve deploy config.yaml
2025-11-03 21:04:51,252 INFO scripts.py:243 -- Deploying from config file: 'config.yaml'.
Traceback (most recent call last):
  File "~/ray/.venv/bin/serve", line 7, in <module>
    sys.exit(cli())
  File "~/ray/.venv/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "~/ray/.venv/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "~/ray/.venv/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "~/ray/.venv/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "~/ray/.venv/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "~/ray/python/ray/serve/scripts.py", line 360, in deploy
    ServeSubmissionClient(address).deploy_applications(
  File "~/ray/python/ray/dashboard/modules/serve/sdk.py", line 80, in deploy_applications
    self._raise_error(response)
  File "~/ray/python/ray/dashboard/modules/dashboard_sdk.py", line 283, in _raise_error
    raise RuntimeError(
RuntimeError: Request failed with status code 500: Traceback (most recent call last):
  File "~/ray/python/ray/dashboard/optional_utils.py", line 188, in decorator
    return await f(self, *args, **kwargs)
  File "~/ray/python/ray/dashboard/modules/serve/serve_head.py", line 37, in check
    return await func(self, *args, **kwargs)
  File "~/ray/python/ray/dashboard/modules/serve/serve_head.py", line 167, in put_all_applications
    client = await serve_start_async(
  File "~/ray/python/ray/serve/_private/api.py", line 148, in serve_start_async
    _check_http_options(client.http_config, http_options)
  File "~/ray/python/ray/serve/_private/api.py", line 51, in _check_http_options
    raise RayServeConfigException(
ray.serve.exceptions.RayServeConfigException: Attempt to update `http_options` or `proxy_location` has been detected! Attempted updates: `{'location': {'previous': 'EveryNode', 'new': 'HeadOnly'}, 'host': {'previous': '0.0.0.0', 'new': '0.0.0.1'}, 'port': {'previous': 8000, 'new': 8001}}`. HTTP config is global to your Ray cluster, and you can't update it during runtime. Please restart Ray Serve to apply the change.

```

The same behavior for serve REST API:
```
bash$ curl -X PUT "http://localhost:8265/api/serve/applications/"   -H "Accept: application/json"   -H "Content-Type: application/json"   -d '{
> "http_options": {
> 
Makefile             config.yaml          config_changed.yaml  discrepancy.py       hello_world_2.py     out                  test_failure.py      
__pycache__/         config_3_apps.yaml   curl                 hello_world.py       new                  rez.json             
> "host": "0.0.0.1"
> },
>   "applications": [
>     {
>       "name": "app1",
>       "route_prefix": "/",
>       "import_path": "hello_world:hello_world_app",
>       "runtime_env": {},
>       "deployments": [
>         { "name": "hello_world" }
>       ]
>     }
>   ]
> }'
Traceback (most recent call last):
  File "~/ray/python/ray/dashboard/optional_utils.py", line 188, in decorator
    return await f(self, *args, **kwargs)
  File "~/ray/python/ray/dashboard/modules/serve/serve_head.py", line 37, in check
    return await func(self, *args, **kwargs)
  File "~/ray/python/ray/dashboard/modules/serve/serve_head.py", line 167, in put_all_applications
    client = await serve_start_async(
  File "~/ray/python/ray/serve/_private/api.py", line 148, in serve_start_async
    _check_http_options(client.http_config, http_options)
  File "~/ray/python/ray/serve/_private/api.py", line 51, in _check_http_options
    raise RayServeConfigException(
ray.serve.exceptions.RayServeConfigException: Attempt to update `http_options` or `proxy_location` has been detected! Attempted updates: `{'host': {'previous': '0.0.0.0', 'new': '0.0.0.1'}}`. HTTP config is global to your Ray cluster, and you can't update it during runtime. Please restart Ray Serve to apply the change.
```

--------------------------------------------

A thing worth to mention is that, this change makes explicit the discrepancy between default `host` value in [serve.config.HTTPOptions](https://github.com/ray-project/ray/blob/master/python/ray/serve/config.py#L433) (host="127.0.0.1") vs [serve.schema.HTTPOptionsSchema](https://github.com/ray-project/ray/blob/master/python/ray/serve/schema.py#L683) (host="0.0.0.0"). `serve.config.HTTPOptions` is primarily used in imperative serve API (Python API or CLI with params) and `serve.schema.HTTPOptionsSchema` is used in declarative serve API (REST API or `deploy/run` with config file)
Previously, when users use commands `start` and then `deploy` or `run` with default params - `http_options` from the `start` command were used.
Now we explicitly failing in this scenario:
```
ray stop
ray start --head
serve build -o config.yaml hello_world:hello_world_app  # generate `config.yaml` with default values

serve start
serve deploy config.yaml
...
ray.serve.exceptions.RayServeConfigException: Attempt to update `http_options` or `proxy_location` has been detected! Attempted updates: `{'host': {'previous': '127.0.0.1', 'new': '0.0.0.0'}}`. HTTP config is global to your Ray cluster, and you can't update it during runtime. Please restart Ray Serve to apply the change.
``` 
Maybe the `host` default value should be aligned in `HTTPOptions` and `HTTPOptionsSchema`.


<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

Closes #56163

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Disallow runtime updates to global `http_options`/`proxy_location` by raising `RayServeConfigException`, fix default proxy location handling, refactor validation, and update docs/tests/CLI accordingly.
> 
> - **Serve API (backend)**
>   - Raise `RayServeConfigException` on attempts to change global `http_options`/`proxy_location` via `_check_http_options` (now compares against `client.http_config`, normalizes `ProxyLocation`/`DeploymentMode`).
>   - Fix default proxy location handling: `_prepare_http_options` sets `DeploymentMode.EveryNode` when `proxy_location=None`; `start()` uses it.
>   - Adjust `serve_start`/`serve_start_async` to pass `client.http_config` into `_check_http_options`.
>   - Add `RayServeConfigException` in `ray.serve.exceptions`.
> - **Dashboard REST**
>   - Remove `validate_http_options` warning logic from `serve_head.py`; rely on backend check.
>   - PUT with changed HTTP/proxy config now fails (500) instead of warning.
> - **CLI/Tests**
>   - Update tests to expect failure on config changes and to use explicit HTTP host where needed.
>   - Add tests for `_prepare_http_options` and `serve.start` rejection on changed HTTP config; add CLI test verifying detailed diff in error.
> - **Docs**
>   - Note that `proxy_location` and HTTP/gRPC configs are cluster-global and cannot be updated at runtime.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 1f690a6ef8db9284d1df246bcc6616dda105cf50. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

axreldable committed:
[serve] Fail on the change of 'proxy_location' and 'http_options' parameters for the 'serve' API

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable commented:
/gemini review

axreldable commented:
/gemini review

axreldable commented on python/ray/serve/scripts.py:
```diff
@@ -269,6 +270,62 @@ def _generate_config_from_file_or_import_path(
+
+    curr_proxy_location = curr_serve_details.proxy_location
+    new_proxy_location = new_config.proxy_location
+    if curr_proxy_location is not None and new_proxy_location != curr_proxy_location:
+        raise RayServeException(
+            f"Attempt to update `proxy_location` from `{curr_proxy_location}` "
+            f"to `{new_proxy_location}` has been detected! "
+            "Proxy config is global to your Ray cluster, and you can't update it during runtime. "
+            "Please restart Ray Serve to apply the change."
+        )
```
Will add this.

axreldable commented:
/gemini review

harshit-anyscale commented on doc/source/serve/production-guide/config.md:
```diff
@@ -60,6 +60,8 @@ The `proxy_location` field configures where to run proxies to handle traffic to
 - HeadOnly: Only run a single proxy on the head node.
 - Disabled: Don't run proxies at all. Set this value if you are only making calls to your applications using deployment handles.
 
+Note that the Proxy config is global to your Ray cluster, and you can't update it during runtime.
```
we should add it for HTTP config as well

harshit-anyscale commented on python/ray/serve/scripts.py:
```diff
@@ -353,7 +414,11 @@ def deploy(
     )
 
-    ServeSubmissionClient(address).deploy_applications(
+    submission_client = ServeSubmissionClient(address)
+    serve_details = ServeInstanceDetails(**submission_client.get_serve_details())
+    _validate_deployment_config(serve_details, config)
+
+    submission_client.deploy_applications(
```
if we make change at this layer, then still PUT API call will be ignoring the warning errors(this change works for serve deploy only), shouldn't we put this validation behind the API? as soon as we intercept the PUT request

harshit-anyscale replied:
first option seems appropriate, as we are already performing other validations there
@abrarsheikh let us know if you think otherwise

axreldable replied:
Hi @harshit-anyscale , @abrarsheikh !

It turns out that it's a bit trickier than I thought. There are 3 places where users can set/update `http_options` now:
- Python API
- CLI
- REST API
```
Python API:
serve:
- start -> _private_api.serve_start
- run -> _private_api.serve_start

Command Line Interface (CLI)
serve start -> serve.start
serve deploy -> REST API PUT "/api/serve/applications/" 
serve run -> _private_api.serve_start

Serve REST API
PUT "/api/serve/applications/" -> _private_api.serve_start_async
```
All of those methods end up in the `_private_api.serve_start(s)` methods and I believe the `_check_http_options(client, http_options)` function is the correct place to introduce this change.

The problem now is that, the info whether `http_options.location` was set by user or not is lost for some branches as that value is manually set based on `proxy_location`. I need to refactor the API a bit to be able to distinguish what user entered and verify the difference.

But while I was looking into it, I identified a bug and a discrepancy/bug which relate to `proxy_location` parameter. It would be good to resolve them first to reduce the complexity of this change. I opened a separate PR #56993 for it.

axreldable replied:
It's finally ready for review. 

@harshit-anyscale , could you please take a look?

harshit-anyscale commented on python/ray/serve/scripts.py:
```diff
@@ -269,6 +270,66 @@ def _generate_config_from_file_or_import_path(
+
+    curr_http_options = curr_serve_details.http_options
+    new_http_options = new_config.http_options
+    if curr_http_options is not None:
+        diff_http_options = {}
+        for option, new_value in new_http_options.dict(exclude_unset=True).items():
+            prev_value = getattr(curr_http_options, option)
+            if prev_value != new_value:
+                diff_http_options[option] = {"previous": prev_value, "new": new_value}
+        if diff_http_options:
+            raise RayServeException(
+                "Attempt to update `http_options` has been detected! "
+                f"Attempted updates: {diff_http_options}. "
+                "HTTP config is global to your Ray cluster, and you can't update it during runtime. "
+                "Please restart Ray Serve to apply the change."
+            )
```
if curr_http_options is None, and someone try to set them for the first time via serve deploy, shouldn't that throw error too?

axreldable replied:
@harshit-anyscale , it turns out that `curr_http_options is None` means that `serve deploy` commands is triggered for the first time. So, no error is needed for it. I've updated the logic to make it easier to read and added comments.

axreldable commented:
Hi @harshit-anyscale , I addressed comments and asked a [question](https://github.com/ray-project/ray/pull/56507#discussion_r2362943150). Could you kindly take a look?

axreldable commented on python/ray/serve/api.py:
```diff
@@ -64,6 +64,24 @@
+        http_options.location = ProxyLocation._to_deployment_mode(proxy_location)
+
+    return http_options
+
```
> The _prepare_http_options function introduces two regressions. It now unconditionally sets http_options.location to DeploymentMode.EveryNode when proxy_location is None, overriding user-provided values. 

Match original behavior.

`string-to-ProxyLocation` is processed in the `ProxyLocation._to_deployment_mode(proxy_location)` function.

jjyao reviewed (approved):
CORE stamp.

jjyao commented:
> Thank the approval. This PR is still WIP. Could you please clarify what CORE stamp. means?

Means core side change looks good and approved.

axreldable committed:
[serve] Fix test: remove serve.start because of different default host in HTTPOptions and HTTPOptionsSchema

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Change default HTTPOptions.location HeadOnly -> EveryNode

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Remove unused tests + change condition

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Remove unused tests + change condition

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Use correct host in test

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Use correct port in test

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Add RayServeConfigException to docs

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Add todo

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Add tests

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable committed:
[serve] Check valid keys

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable commented on python/ray/serve/tests/test_telemetry_1.py:
```diff
@@ -149,7 +150,7 @@ def test_rest_api(manage_ray_with_telemetry, tmp_dir):
 
     # Check that app deletions are tracked.
-    new_config = {"applications": []}
+    new_config = {"http_options": {"host": "127.0.0.1"}, "applications": []}
```
There is a discrepancy between default host value in [serve.config.HTTPOptions](https://github.com/ray-project/ray/blob/master/python/ray/serve/config.py#L433) (host="127.0.0.1") vs [serve.schema.HTTPOptionsSchema](https://github.com/ray-project/ray/blob/master/python/ray/serve/schema.py#L683) (host="0.0.0.0").

I tried to summarize it in the description.

<img width="968" height="578" alt="image" src="https://github.com/user-attachments/assets/09c87e3c-dafa-44d7-bf2a-7b12d51a3bc9" />

axreldable commented on python/ray/serve/config.py:
```diff
@@ -584,7 +584,7 @@ class HTTPOptions(BaseModel):
     port: int = DEFAULT_HTTP_PORT
     middlewares: List[Any] = []
-    location: Optional[DeploymentMode] = DeploymentMode.HeadOnly
+    location: Optional[DeploymentMode] = DeploymentMode.EveryNode
```
@abrarsheikh , please also notice that I change this default value. Aiming to align it with Proxy-location default. Without this if I remember correctly it's challenging to align all of the tests and can be confusing for users either.

axreldable committed:
[serve] Update docstring

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

harshit-anyscale reviewed (approved):
apart from https://github.com/ray-project/ray/pull/56507#discussion_r2514931762, else LGTM!

axreldable committed:
Merge branch 'master' into 56163_proxy_location
